firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -0,0 +1,277 @@
+      (runnable, executor) -> {
+        // We ignore rejected executions, since some gRPC messages can arrive after shutdown.
+      };
+
+  private static Supplier<ManagedChannelBuilder<?>> overrideChannelBuilderSupplier;
+
+  private final ExecutorService channelQueue;
+  private ManagedChannel channel;
+  private CallOptions callOptions;
+  private boolean shutdown = false;
+
+  /**
+   * Helper function to globally override the channel that RPCs use. Useful for testing when you
+   * want to bypass SSL certificate checking.
+   *
+   * @param channelBuilderSupplier The supplier for a channel builder that is used to create gRPC
+   *     channels.
+   */
+  @VisibleForTesting
+  public static void overrideChannelBuilder(
+      Supplier<ManagedChannelBuilder<?>> channelBuilderSupplier) {
+    overrideChannelBuilderSupplier = channelBuilderSupplier;
+  }
+
+  GrpcCallProvider(
+      AsyncQueue workerQueue,
+      Context context,
+      DatabaseInfo databaseInfo,
+      CallCredentials firestoreHeaders) {
+    ThreadFactory threadFactory =
+        runnable -> {
+          Thread thread = Executors.defaultThreadFactory().newThread(runnable);
+          thread.setName("FirestoreGrpcWorker");
+          thread.setDaemon(true);
+          thread.setUncaughtExceptionHandler(
+              (crashingThread, throwable) -> workerQueue.panic(throwable));
+          return thread;
+        };
+
+    this.channelQueue =
+        new ThreadPoolExecutor(
+            /* corePoolSize= */ 1,
+            /* maximumPoolSize= */ 1,
+            /* keepAliveTime= */ 0L,
+            TimeUnit.MILLISECONDS,
+            new LinkedBlockingQueue<>(),
+            threadFactory,
+            IGNORE_REJECTIONS_HANDLER);
+
+    // We execute network initialization on an internal queue to not block operations that
+    // depend on the AsyncQueue.
+    channelQueue.execute(
+        () -> {
+          channel = initChannel(workerQueue, context, databaseInfo);
+          FirestoreGrpc.FirestoreStub firestoreStub =
+              FirestoreGrpc.newStub(channel).withCallCredentials(firestoreHeaders);
+          callOptions = firestoreStub.getCallOptions();
+        });
+  }
+
+  /** Sets up the SSL provider and configures the gRPC channel. */
+  private ManagedChannel initChannel(
+      AsyncQueue workerQueue, Context context, DatabaseInfo databaseInfo) {
+    try {
+      // We need to upgrade the Security Provider before any network channels are initialized.
+      // `OkHttp` maintains a list of supported providers that is initialized when the JVM first
+      // resolves the static dependencies of ManagedChannel.
+      ProviderInstaller.installIfNeeded(context);
+    } catch (GooglePlayServicesNotAvailableException /* Thrown by ProviderInstaller */
+        | GooglePlayServicesRepairableException /* Thrown by ProviderInstaller */
+        | IllegalStateException e /* Thrown by Robolectric */) {
```

Comment:
Note that we can get rid of this if we change MockDatastore to not call the main Datastore constructor, which right now always initializes GrpcCallProvider.